Feature/validations model specs and n+1

leolpaz opened the pull request:
### Ruby/Blog-App

Added on this pull request:

- User, Posts, Likes, Comments
  - Create validations for models
  - Create tests
  - Views for posts and comments
  - Flash messages to all create methods
    - Integration with views

leolpaz committed:
Create user validation

leolpaz committed:
Create post validation

leolpaz committed:
Write user tests and fix a bug with user posts ordering

leolpaz committed:
Write tests for posts and fix a bug with validation invalidating 0 as a counter

leolpaz committed:
Write tests for comments and likes

leolpaz committed:
Solve found instances of n+1 problems

leolpaz committed:
Update readme

leolpaz committed:
Fix formatting issues pointed by Rubocop

leolpaz committed:
Fix request tests

leolpaz committed:
Add flash messages

Lameck1 reviewed (changes requested):
Hi, @leolpaz, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

#### Since you did not add code in the `index` action in `app/controllers/posts_controller.rb`, it is not trackable and therefore I cannot leave an inline comment. The change you are supposed to work on will be left here in the review summary.

- [ ] Please let's ensure we take care of `N + 1` problems in the `index` action of the posts controller.

```ruby
# Besides finding the user, let's also eager load all the posts along with associated comments in the index action. I have seen you've tried doing this in the post model, I would like to encourage you to do it in more rails way and handle the logic of all the data we need in the posts controller.

def index
  @user = User.find(params[:user_id])
  @posts = @user.posts.includes(:comments)
end

# [OPTIONAL] This means that you will need to update the code in `app/views/posts/index.html.erb`, the code for rendering all posts will look like what we have below:

# Instead of @user.posts.each, now do @posts.each since the controller has fetched all the posts for you. It is not a good practice to let the views do the fetching of data.

<% @posts.each do |post| %>
  <%= link_to [post.user, post] do %></h2>
    <div>
      <h3><%= post.title %></h3>
      <p><%= post.text %></p>
      <span>Comments: <%= post.comments_counter %>, Likes: <%= post.likes_counter %></span>
    </div>
    <% post.comments.each do |comment| %>
      <p><%= comment.user.name %>: <%= comment.text %></p>
    <% end %>
  <% end %>
<% end %>
```
- [ ] Checking comments under the review NOT APPLICABLE. All the best.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

leolpaz committed:
Implement changes on the posts_controller and view as suggested by reviewer

leolpaz commented:
Thank you so much for the detailed review, I implemented the suggested changes, but also added a limit of 5 to @posts on the index, as according to the wireframe only 5 posts should be shown, again thank you so much for the explanation @Lameck1 .

uwadonat reviewed (approved):
Hi @leolpaz, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Status: Approved ✔️✔️✔️

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```
